examples/matching_engine/src/contract.rs

```diff
@@ -0,0 +1,640 @@
+        let mut iter = view.queue.iter_mut().await?;
+        let mut remove_order = Vec::new();
+        loop {
+            let order = iter.next();
+            match order {
+                Some(order) => {
+                    let fill = min(order.count, *count);
+                    count.try_sub_assign(fill).unwrap();
+                    order.count.try_sub_assign(fill).unwrap();
+                    transfers.extend_from_slice(&Self::get_transfers(
+                        nature,
+                        fill,
+                        owner,
+                        order,
+                        price_level,
+                        price_insert,
+                    ));
+                    if order.count == Amount::zero() {
+                        remove_order.push((order.owner, order.order_id));
+                    }
+                    if *count == Amount::zero() {
+                        break;
+                    }
+                }
+                None => {
+                    break;
+                }
+            }
+        }
+        Self::remove_zero_orders_from_level(view).await?;
+        Ok(remove_order)
+    }
+
+    async fn insert_order(
+        &mut self,
+        owner: AccountOwner,
+        nature: OrderNature,
+        order_id: OrderId,
+        price: Price,
+    ) -> Result<(), MatchingEngineError> {
+        let account_info = self.account_info.get_mut_or_default(&owner).await?;
+        account_info.orders.insert(order_id);
+        let key_book = KeyBook {
+            price,
+            nature,
+            owner,
+        };
+        self.orders.insert(&order_id, key_book)?;
+        Ok(())
+    }
+
+    async fn remove_order_id(
+        &mut self,
+        entry: (AccountOwner, OrderId),
+    ) -> Result<(), MatchingEngineError> {
+        let (owner, order_id) = entry;
+        let account_info = self
+            .account_info
+            .get_mut(&owner)
+            .await
+            .expect("account_info")
+            .unwrap();
+        account_info.orders.remove(&order_id);
+        Ok(())
+    }
+
+    async fn remove_order_ids(
+        &mut self,
+        entries: Vec<(AccountOwner, OrderId)>,
+    ) -> Result<(), MatchingEngineError> {
+        for entry in entries {
+            self.remove_order_id(entry).await?;
+        }
+        Ok(())
+    }
+
+    async fn insert_and_uncross_market(
+        &mut self,
+        owner: &AccountOwner,
+        count: Amount,
+        nature: OrderNature,
+        price: &Price,
+    ) -> Result<Vec<Transfer>, MatchingEngineError> {
+        // Bids are ordered from the highest bid (most preferable) to the smallest bid.
+        // Asks are ordered from the smallest (most preferable) to the highest.
+        // The price are ordered by having a custom ordering. Under this natural ordering
+        // we go from small to big.
+        // So, we keep the ask price in the same order but we revert the order of the bid prices.
```

Comment:
```suggestion
        // The prices have custom serialization so that they are in increasing order.
        // To reverse the order of the bids, we take the bitwise complement of the price.
```